### ISSUE TITLE
制作landmark数据的时候，在计算loss的时候，能否把这些图片也当成正例来计算loss？

### ISSUE THREAD

**Issue by xiaoxinyi** [AITTSMD/MTCNN-Tensorflow#22]
在使用`gen_landmark_12_aug.py`生成landmark数据时，把label设成了-2。

```
f.write(join(dstdir, "%d.jpg" % (image_id)) + " -2 " + " ".join(landmarks) + "\n")
```

能不能把-2的也在难例计算loss的时候当作正例来计算?

因为在做ohem的时候，函数`def cls_ohem(cls_prob, label):`中，

```
...
label_filter_invalid = tf.where(tf.less(label, 0), zeros, label)
num_cls_prob = tf.size(cls_prob)
# 把cls_prob_reshape展开成batch*2 X 1
cls_prob_reshape = tf.reshape(cls_prob, [num_cls_prob, -1])
label_int = tf.cast(label_filter_invalid, tf.int32)
num_row = tf.to_int32(cls_prob.get_shape()[0])
row = tf.range(num_row) * 2
indices_ = row + label_int
label_prob = tf.squeeze(tf.gather(cls_prob_reshape, indices_))
loss = -tf.log(label_prob + 1e-10)
```

把landmark的所有脸，也都变成为负例来计算loss。

而在选择难例的时候，会把之前作为负例计算的loss，全部筛掉。

```
    zeros = tf.zeros_like(label_prob, dtype=tf.float32)
    ones = tf.ones_like(label_prob, dtype=tf.float32)
    valid_inds = tf.where(label < zeros, zeros, ones)
    num_valid = tf.reduce_sum(valid_inds)
    keep_num = tf.cast(num_valid * num_keep_radio, dtype=tf.int32)
    # set 0 to invalid sample
    loss = loss * valid_inds
    loss, _ = tf.nn.top_k(loss, k=keep_num)
```

为什么在挖掘难例的时候，只挖正例， 而不对负例进行挖掘？

**Comment by AITTSMD** [comment]
@xiaoxinyi 
1 cls_ohem仅用来计算分类(正负样本)的loss，与landmark部分无关。landmark部分的数据集虽然也有人脸位置的标签，但我没有使用。
2 难例挖掘的时候，是对所有样本的loss值从大到小排序，选择topK，并不是只针对正例挖掘。

**Comment by xiaoxinyi** [comment]
@AITTSMD 

```
# label 如果小于 0， 那么valid_inds对应的位置也取0，
# label 如果等于 1，那么valid_inds对应的位置取1.
valid_inds = tf.where(label < zeros, zeros, ones)
# valid_inds为0的位置，与loss相乘之后也变成0
# valid_inds为1的位置，loss值保留
loss = loss * valid_inds
```

这两句代码，loss只计算了label == 1, 筛选掉了非正例的loss。所以按照我对这段代码的理解，我觉得只是对正例进行了挖掘。

**Comment by AITTSMD** [comment]
@xiaoxinyi 不好意思，回复晚了。您是对的，这里是我疏忽了，原论文中也是对所有的样本进行难例挖掘。谢谢您的提醒～

**Comment by xiaoxinyi** [comment]
@AITTSMD Thanks.

**Comment by guangdongliang** [comment]
@xiaoxinyi @AITTSMD 
part,neg,pos对应label分别为-1,0,1,因此如下代码只是将part样本筛选掉了吧.
    valid_inds = tf.where(label < zeros, zeros, ones)

**Comment by JinbiaoYu** [comment]
@AITTSMD 这段代码只计算了label=1的loss吗，label=0的loss没有计算吗

**Comment by yeyupiaoling** [comment]
@AITTSMD @xiaoxinyi @guangdongliang 
请问在cls_ohem是不是已经把landmark数据当作成没有人脸数据来处理了呢？landmark数据也是有完整的人脸呢？望解答。

https://github.com/AITTSMD/MTCNN-Tensorflow/blob/3b3934d38f8d34287cc933a581537a1acfd0bb60/train_models/mtcnn_model.py#L24-L29
